internal/painter/gl/draw.go

```diff
@@ -97,12 +100,147 @@ func (p *painter) drawRaster(img *canvas.Raster, pos fyne.Position, frame fyne.S
+
+	rectCoordsUniform := p.ctx.GetUniformLocation(p.rectangleProgram, "rect_coords")
+	x1Scaled := roundToPixel(points[0]*p.pixScale, 1.0)
+	x2Scaled := roundToPixel(points[4]*p.pixScale, 1.0)
+	y1Scaled := roundToPixel(points[1]*p.pixScale, 1.0)
+	y2Scaled := roundToPixel(points[9]*p.pixScale, 1.0)
+	p.ctx.Uniform4f(rectCoordsUniform, x1Scaled, x2Scaled, y1Scaled, y2Scaled)
+
+	strokeUniform := p.ctx.GetUniformLocation(p.rectangleProgram, "stroke_width")
+	strokeWidthScaled := roundToPixel(rect.StrokeWidth*p.pixScale, 1.0)
+	p.ctx.Uniform1f(strokeUniform, strokeWidthScaled)
+
+	fillColorUniform := p.ctx.GetUniformLocation(p.rectangleProgram, "fill_color")
+	rF, gF, bF, aF := rect.FillColor.RGBA()
+	if aF == 0 {
+		p.ctx.Uniform4f(fillColorUniform, 0, 0, 0, 0)
+	} else {
+		alphaF := float32(aF)
+		colF := []float32{float32(rF) / alphaF, float32(gF) / alphaF, float32(bF) / alphaF, alphaF / 0xffff}
+		p.ctx.Uniform4f(fillColorUniform, colF[0], colF[1], colF[2], colF[3])
+	}
+
+	strokeColorUniform := p.ctx.GetUniformLocation(p.rectangleProgram, "stroke_color")
+	var col color.Color
+	if rect.StrokeColor == col {
+		rect.StrokeColor = color.NRGBA{0.0, 0.0, 0.0, 0.0}
 	}
-	p.drawTextureWithDetails(rect, p.newGlRectTexture, pos, rect.Size(), frame, canvas.ImageFillStretch,
-		1.0, paint.VectorPad(rect))
+	rS, gS, bS, aS := rect.StrokeColor.RGBA()
+	if aS == 0 {
+		p.ctx.Uniform4f(strokeColorUniform, 0, 0, 0, 0)
+	} else {
+		alphaS := float32(aS)
+		colF := []float32{float32(rS) / alphaS, float32(gS) / alphaS, float32(bS) / alphaS, alphaS / 0xffff}
+		p.ctx.Uniform4f(strokeColorUniform, colF[0], colF[1], colF[2], colF[3])
+	}
+	p.logError()
+	// Fragment: END
+
+	p.ctx.DrawArrays(triangles, 0, 6)
+	p.logError()
+	p.freeBuffer(vbo)
+}
+
+func (p *painter) drawRoundRectangle(
+	rect *canvas.Rectangle,
+	pos fyne.Position,
+	frame fyne.Size,
+) {
+	// Vertex: BEG
+	points := p.vecRectCoords(pos, rect, frame)
+	p.ctx.UseProgram(p.roundRectangleProgram)
+	vbo := p.createBuffer(points)
+	p.defineVertexArray(p.roundRectangleProgram, "vert", 2, 4, 0)
+	p.defineVertexArray(p.roundRectangleProgram, "normal", 2, 4, 2)
+
+	p.ctx.BlendFunc(srcAlpha, oneMinusSrcAlpha)
+	p.logError()
+	// Vertex: END
+
+	// Fragment: BEG
+	frameSizeUniform := p.ctx.GetUniformLocation(p.roundRectangleProgram, "frame_size")
+	frameWidthScaled := roundToPixel(frame.Width*p.pixScale, 1.0)
+	frameHeightScaled := roundToPixel(frame.Height*p.pixScale, 1.0)
+	p.ctx.Uniform4f(frameSizeUniform, frameWidthScaled, frameHeightScaled, 0.0, 0.0)
+
+	rectCoordsUniform := p.ctx.GetUniformLocation(p.roundRectangleProgram, "rect_coords")
+	x1Scaled := roundToPixel(points[0]*p.pixScale, 1.0)
+	x2Scaled := roundToPixel(points[4]*p.pixScale, 1.0)
+	y1Scaled := roundToPixel(points[1]*p.pixScale, 1.0)
+	y2Scaled := roundToPixel(points[9]*p.pixScale, 1.0)
+	p.ctx.Uniform4f(rectCoordsUniform, x1Scaled, x2Scaled, y1Scaled, y2Scaled)
```

Comment:
There is a bit of duplicated code between the initialization for both shader, it would be great to refactor the common code as much as possible.

Reply:
3 new func, code reuse - done